server/src/test/java/org/elasticsearch/cluster/ClusterStateIT.java

```diff
@@ -0,0 +1,351 @@
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TcpTransport;
+import org.elasticsearch.watcher.ResourceWatcherService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.gateway.GatewayService.STATE_NOT_RECOVERED_BLOCK;
+import static org.elasticsearch.test.ESIntegTestCase.Scope.TEST;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.not;
+import static org.hamcrest.Matchers.hasItem;
+import static org.hamcrest.Matchers.instanceOf;
+
+/**
+ * This test suite sets up a situation where the cluster has two plugins installed (node, and node-and-transport-client), and a transport
+ * client only has node-and-transport-client plugin installed. Each of these plugins inject customs into the cluster state and we want to
+ * check that the client can de-serialize a cluster state response based on the fact that the response should not contain customs that the
+ * transport client does not understand based on the fact that it only presents the node-and-transport-client-feature.
+ */
+@ESIntegTestCase.ClusterScope(scope = TEST)
+public class ClusterStateIT extends ESIntegTestCase {
+
+    public abstract static class Custom implements MetaData.Custom {
+
+        private static final ParseField VALUE = new ParseField("value");
+
+        private final int value;
+
+        int value() {
+            return value;
+        }
+
+        Custom(final int value) {
+            this.value = value;
+        }
+
+        Custom(final StreamInput in) throws IOException {
+            value = in.readInt();
+        }
+
+        @Override
+        public EnumSet<MetaData.XContentContext> context() {
+            return MetaData.ALL_CONTEXTS;
+        }
+
+        @Override
+        public Diff<MetaData.Custom> diff(final MetaData.Custom previousState) {
+            return null;
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            out.writeInt(value);
+        }
+
+        public static int fromXContent(final XContentParser parser) throws IOException {
+            XContentParser.Token token;
+            int value = 0;
+            String currentFieldName = null;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                switch (token) {
+                    case FIELD_NAME:
+                        currentFieldName = parser.currentName();
+                        break;
+                    case VALUE_BOOLEAN:
```

Comment:
I think this method does not need to be implemented for this test. It is buggy anyhow, I think, as this should be `VALUE_NUMBER`

Reply:
I pushed 66f2ea95dc29438b1772321fc94afd7e13cae9d7.